go/vt/vttablet/tabletserver/vstreamer/helper_event_test.go

```diff
@@ -0,0 +1,569 @@
+		return err
+	}
+	ts.fieldEvents = make(map[string]*TestFieldEvent)
+	ts.fieldEventsSent = make(map[string]bool)
+	ts.state = make(map[string]*query.Row)
+	ts.metadata = make(map[string][]string)
+	ts.pkColumns = make(map[string][]string)
+	// create tables
+	for i, t := range ts.schema.Tables() {
+		execStatement(ts.t, ts.ddls[i])
+		fe := ts.getFieldEvent(t)
+		ts.fieldEvents[t.Name()] = fe
+
+		var pkColumns []string
+		var hasPK bool
+		for _, index := range t.TableSpec.Indexes {
+			if index.Info.Type == sqlparser.IndexTypePrimary {
+				for _, col := range index.Columns {
+					pkColumns = append(pkColumns, col.Column.String())
+				}
+				hasPK = true
+			}
+		}
+		if !hasPK {
+			// add all columns as pk columns
+			for _, col := range t.TableSpec.Columns {
+				pkColumns = append(pkColumns, col.Name.String())
+			}
+		}
+		ts.pkColumns[t.Name()] = pkColumns
+	}
+	engine.se.Reload(context.Background())
+	return nil
+}
+
+// Close() should be called (via defer) at the end of the test to clean up the tables created in the test.
+func (ts *TestSpec) Close() {
+	dropStatement := fmt.Sprintf("drop tables %s", strings.Join(ts.schema.TableNames(), ", "))
+	execStatement(ts.t, dropStatement)
+}
+
+func (ts *TestSpec) getBindVarsForInsert(stmt sqlparser.Statement) (string, map[string]string) {
+	bv := make(map[string]string)
+	ins := stmt.(*sqlparser.Insert)
+	tn, _ := ins.Table.TableName()
+	table := tn.Name.String()
+	fe := ts.fieldEvents[table]
+	vals, _ := ins.Rows.(sqlparser.Values)
+	for _, val := range vals {
+		for i, v := range val {
+			bufV := sqlparser.NewTrackedBuffer(nil)
+			v.Format(bufV)
+			s := bufV.String()
+			switch fe.cols[i].dataTypeLowered {
+			case "varchar", "char", "binary", "varbinary", "blob", "text":
+				s = strings.Trim(s, "'")
+			case "set", "enum":
+				s = ts.getMetadataMap(table, fe.cols[i], s)
+			}
+			bv[fe.cols[i].name] = s
+		}
+	}
+	return table, bv
+}
+
+func (ts *TestSpec) getBindVarsForUpdate(stmt sqlparser.Statement) (string, map[string]string) {
+	bv := make(map[string]string)
+	upd := stmt.(*sqlparser.Update)
+	buf := sqlparser.NewTrackedBuffer(nil)
+	upd.TableExprs[0].(*sqlparser.AliasedTableExpr).Expr.Format(buf)
+	table := buf.String()
+	fe, ok := ts.fieldEvents[table]
+	require.True(ts.t, ok, "field event for table %s not found", table)
+	index := int64(0)
+	state := ts.getCurrentState(table)
+	for i, col := range fe.cols {
+		bv[col.name] = string(state.Values[index : index+state.Lengths[i]])
+		index += state.Lengths[i]
+	}
+	for _, expr := range upd.Exprs {
+		bufV := sqlparser.NewTrackedBuffer(nil)
+		bufN := sqlparser.NewTrackedBuffer(nil)
+		expr.Expr.Format(bufV)
+		expr.Name.Format(bufN)
+		bv[bufN.String()] = strings.Trim(bufV.String(), "'")
+	}
+	return table, bv
+}
+
+// Run() runs the test. It first initializes the test, then runs the queries and validates the events.
+func (ts *TestSpec) Run() {
+	require.NoError(ts.t, engine.se.Reload(context.Background()))
+	if !ts.inited {
```

Comment:
This isn't concurrency safe, if it matters. Not sure that it does today. If it does, however, we can use `atomic.Bool` and only set it to true when we've successfully inited.

Reply:
can also add a `sync.Once` to the testspec struct and do `ts.initOnce.Do(func() { ... })`

Reply:
Currently none of the vstreamer tests are concurrency safe. They instantiate a singleton engine.  We will need to refactor, if we want to support running in parallel.